### ISSUE TITLE
Unable to construct GP model from log file

### ISSUE THREAD

**Issue by owaisahmad18** [mir-group/flare#291]
After parsing log file for copper, i am unable to construct GP model. Kindly help.

![image](https://user-images.githubusercontent.com/38828004/131702685-7bde5c58-079f-4a55-90ec-b6529a699a2c.png)

**Comment by YuuuXie** [comment]
Hi @owaisahmad18 ,

Based on the error information, you are parsing a restarted OTF log file. To build a complete GP, you need to parse the first log file (before restarting), to get an initial GP `init_gp`. And then parse the second log file (for the restarted OTF). The final complete GP model can be obtained by
```
gp_model = otf_object.make_gp(hyp_no=hyp_no, init_gp=init_gp)
```

**Comment by YuuuXie** [comment]
Hi @owaisahmad18 
It seems that the error is because the two numbers are written without space in between. Could you please attach your out file?

To parse the 2nd file, just call `OtfAnalysis` for the 2nd time
```
otf_object_1 = OtfAnalysis(file1)
init_gp = otf_object_1.make_gp(...)

otf_object_2 = OtfAnalysis(file2)
gp_model = otf_object_2.make_gp(..., init_gp=init_gp)
```

**Comment by owaisahmad18** [comment]
[Cu_otf.out.txt](https://github.com/mir-group/flare/files/7100574/Cu_otf.out.txt)

**Comment by YuuuXie** [comment]
Hi @owaisahmad18 
Looking at the out file, the force error seems to be very large, 
```
mean absolute error: 17.5775 eV/A
mean absolute dft component: 0.0126 eV/A
```
this might be because your initial structure (frame 0) is highly symmetric. Then the environments of atom 0, 1, 2, 3 are almost identical, so when GP adds them, the kernel matrix is degenerate. 

Maybe try randomly perturbing the atomic positions of the initial frame (the positions of the arg `super_cell` that inputs to the `ASE_OTF`)
```python
test_otf = ASE_OTF(super_cell, ...)
```
Then try to start training and see if the mean absolute error reported in the log file becomes reasonable

**Comment by owaisahmad18** [comment]
Ok can i reduce the error if i increase the number of training environments by increasing the number of atoms, if yes then which variables should i make changes to, as i want to increase the number of atoms anyway.

**Comment by YuuuXie** [comment]
No, this is not an issue with number of atoms.

If you want to increase the number of atoms, just increase the number of atoms of the initial structure, i.e. the `super_cell` in the input

**Comment by owaisahmad18** [comment]
ok thanks, i don't see any thread for random perturbation in the documentation, also no variable related to random perturbation neither in bulk() or in ASE_otf(). can you please be a little more specific.

**Comment by YuuuXie** [comment]
Just directly modify the positions of the ASE Atoms, below is an example

# Randomly jitter the atoms to give nonzero forces in the first frame.
jitter_factor = 0.1
for atom_pos in super_cell.positions:
    for coord in range(3):
        atom_pos[coord] += (2 * np.random.rand() - 1) * jitter_factor

I will update this to the tutorial

**Comment by owaisahmad18** [comment]
I made the corrections as suggested ,same error... here is the output file
[Cu_otf.out.txt](https://github.com/mir-group/flare/files/7100805/Cu_otf.out.txt)

**Comment by owaisahmad18** [comment]
[Cu.pz-d-rrkjus.UPF.txt](https://github.com/mir-group/flare/files/7100866/Cu.pz-d-rrkjus.UPF.txt)


here is the complete script along with pseudo potential, i really appreciate if you can point out any mistake.

import numpy as np
from ase import units
from ase.spacegroup import crystal
from ase.build import bulk
from ase.visualize import view

np.random.seed(12345)

a = 3.628
super_cell = bulk('Cu','fcc',a=a,cubic=True)
jitter_factor = 0.1
for atom_pos in super_cell.positions:
    for coord in range(3):
        atom_pos[coord] += (2 * np.random.rand() - 1) * jitter_factor
# view(super_cell, viewer="x3d")
from flare.gp import GaussianProcess
from flare.utils.parameter_helper import ParameterHelper

# set up GP hyperparameters
kernels = ['twobody', 'threebody'] # use 2+3 body kernel
parameters = {'cutoff_twobody': 5.0,
              'cutoff_threebody': 3.5}
pm = ParameterHelper(
    kernels = kernels,
    random = True,
    parameters=parameters
)

hm = pm.as_dict()
hyps = hm['hyps']
cut = hm['cutoffs']
print('hyps', hyps)

gp_model = GaussianProcess(
    kernels = kernels,
    component = 'mc', # If you are using ASE, please set to "mc" no matter for single-component or multi-component
    hyps = hyps,
    cutoffs = cut,
    hyp_labels = ['sig2','ls2','sig3','ls3','noise'],
    opt_algorithm = 'L-BFGS-B',
    n_cpus = 1
)
from flare.mgp import MappedGaussianProcess

grid_params = {'twobody':   {'grid_num': [64]},
               'threebody': {'grid_num': [16, 16, 16]}}

mgp_model = MappedGaussianProcess(grid_params,
                                  unique_species = [6],
                                  n_cpus = 1,
                                  var_map=None)
from flare.ase.calculator import FLARE_Calculator

flare_calculator = FLARE_Calculator(gp_model,
                                    par = True,
                                    mgp_model = None,
                                    use_mapping = False)

super_cell.set_calculator(flare_calculator)
import os
from ase.calculators.espresso import Espresso

# ---------------- set up executable ----------------
label = 'Cu'
input_file = label+'.pwi'
output_file = label+'.pwo'
no_cpus = 4
npool = 4
pw_loc = '/home/owais/Downloads/q-e-qe-6.8/PW/src/pw.x'

# serial
# os.environ['ASE_ESPRESSO_COMMAND'] = f'{pw_loc} < {input_file} > {output_file}'

## parallel qe using mpirun
os.environ['ASE_ESPRESSO_COMMAND'] = f'mpirun -np {no_cpus} {pw_loc} -npool {npool} < {input_file} > {output_file}'

## parallel qe using srun (for slurm system)
# os.environ['ASE_ESPRESSO_COMMAND'] = 'srun -n {no_cpus} --mpi=pmi2 {pw_loc} -npool {npool} < {input_file} > {output_file}'


# -------------- set up input parameters --------------
input_data = {'control':   {'prefix': label,
                            'pseudo_dir': './',
                            'outdir': './out',
                            'calculation': 'scf'},
              'system':    {'ibrav': 0 ,
                            'ecutwfc': 30,
                            'ecutrho': 150,
                            'occupations': 'smearing',
                           'degauss': 0.01},
              'electrons': {
                            'electron_maxstep': 100,
                            'mixing_beta': 0.5}}

# ----------------  pseudo-potentials -----------------
ion_pseudo = {'Cu': 'Cu.pz-d-rrkjus.UPF'}

# -------------- create ASE calculator ----------------
dft_calc = Espresso(pseudopotentials=ion_pseudo, label=label,
                    tstress=True, tprnfor=True, nosym=True,
                    input_data=input_data, kpts=(5, 5, 5))
from ase import units
from ase.md.velocitydistribution import (MaxwellBoltzmannDistribution,
                                         Stationary, ZeroRotation)

temperature = 300
MaxwellBoltzmannDistribution(super_cell, temperature * units.kB)
Stationary(super_cell)  # zero linear momentum
ZeroRotation(super_cell)  # zero angular momentum

md_engine = 'VelocityVerlet'
md_kwargs = {}
from flare.ase.otf import ASE_OTF

otf_params = {'init_atoms': [0, 1, 2, 3],
              'output_name': 'Cu_otf',
              'std_tolerance_factor': 2,
              'max_atoms_added' : 4,
              'freeze_hyps': 10,
              'write_model': 3} # If you will probably resume the training, please set to 3

test_otf = ASE_OTF(super_cell,
                   timestep = 1 * units.fs,
                   number_of_steps = 3,
                   dft_calc = dft_calc,
                   md_engine = md_engine,
                   md_kwargs = md_kwargs,
                   **otf_params)

test_otf.run()

**Comment by YuuuXie** [comment]
Could you please try 
```python
test_otf = ASE_OTF(..., force_only=True)
```

It might be a similar issue as https://github.com/mir-group/flare/issues/280#issuecomment-786767130

**Comment by owaisahmad18** [comment]
Thanks, it worked
but i have two doubts:
1. why we need single otf file for AgI while two for lets say Copper.
2.   lammps script run successfully if set copper lattice paramter to 4 ![image](https://user-images.githubusercontent.com/38828004/132130651-94a01e70-ea1c-4a9b-b331-862a233c5dbe.png)
though results are poor but if set the same variable to "3+ ..." then it shows
![image](https://user-images.githubusercontent.com/38828004/132130692-17c12705-4280-40d4-978f-5407500e116a.png)
so how can i change the lower bound values for 2b

**Comment by YuuuXie** [comment]
Hi @owaisahmad18 
1. If your run OTF and it's finished without interruption, then only one file is needed. Maybe your copper training was done/interrupted for one time, and then restarted, so you get two log files, and you need to load data from both of them to get a complete training data set.
2. Please check https://github.com/mir-group/flare/issues/284

**Comment by owaisahmad18** [comment]
Its working now, Thanks a lot for being so helpful and patient.
